api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py

```diff
@@ -640,16 +698,24 @@ def __len__(self) -> int:
     def __getitem__(self, index: int) -> ObsAndXDatum:
         raise NotImplementedError("IterDataPipe can only be iterated")
 
-    def _build_obs_encoders(self, query: soma.ExperimentAxisQuery) -> Encoders:
+    def _build_obs_encoders(self, query: soma.ExperimentAxisQuery) -> List[Encoder]:
         pytorch_logger.debug("Initializing encoders")
 
-        obs = query.obs(column_names=self.obs_column_names).concat()
-        encoders = {}
-        for col in self.obs_column_names:
-            if obs[col].type in (pa.string(), pa.large_string()):
-                enc = LabelEncoder()
-                enc.fit(obs[col].combine_chunks().unique())
-                encoders[col] = enc
+        encoders = []
+        obs = query.obs(column_names=self.obs_column_names).concat().to_pandas()
+
+        if self._custom_encoders:
+            # Register all the custom encoders with obs
+            for enc in self._custom_encoders:
+                enc.register(obs)
+                encoders.append(enc)
+        else:
```

Comment:
⚠️ Should this still register default encoders for columns that are not the source of any custom encoder? This complicates matters, I realize, since each encoder would then have to publicize its source columns explicitly.  But without it, any `object`-typed columns that do not have a custom encoder will not be usable in a Tensor. Minimally, the class should present a clear error if after transform, any non-numeric columns remain (with a hint to specify an encoder for each unencoded column).

Reply:
My idea here is that, if the user specifies custom encoders, then they are responsible for ensuring that all the columns are transformed properly. With the new changes (see lined 346-347), if a column isn't explicitly registered by the encoder, it won't be added to the final obs_tensor. I believe this addresses this concern, but feel free to let me know if I am still missing something 😄 

That said, I think we should expand the docstrings to better explain the consequences of using custom encoders.

Reply:
In that case, I suppose the `obs_columns` and `encoders` could be mutually exclusive. That is, it would be nice if one could specify just the `encoders` without having to also specify the source `obs_columns`, since this could lead to errors.  But here again, the `Encoder` would need to be able to report what columns it depends on so that the `ExpDataPipe` can retrieve them correctly.

But I also agree this could just be handled via clear documentation and useful error messages.

Reply:
I do agree that, if using custom encoders, `obs_columns` could be omitted. There is a slight performance increase in using them, as it doesn't require to fetch the whole `obs` though. Maybe we can just make them optional?

Reply:
My 2 cents is that this should error if both are passed, and the encoders should be able to report which columns they are going to grab. This links onto my other comment about adopting a scikit-learn like (or maybe even compatible) API, where column transformers and metadata routing are used to retrieve information like column names for their transformers.

Reply:
In my heart of hearts, I would probably even allow specifying an encoder for `X`. E.g. instead of specifying `use_sparse_X` you could have a `DenseEncoder()`.

Reply:
I'm ok with implementing the logic where either the parameter can be defined. I also think that using an encoder for X is a great idea - I'll create a ticket for that as it will go out of scope for this PR.

Reply:
looks like the obs_columns vs encoder are now mutually exclusive, so the comment is nearly addressed--does the X encoder idea still need an issue created for it? (didn't see one)

Reply:
Added: https://github.com/chanzuckerberg/cellxgene-census/issues/1219